tools/distrib/python/grpcio_tools/grpc_tools/protoc.py

```diff
@@ -31,6 +36,124 @@ def main(command_arguments):
+                                                     protobuf_path)
+            module = importlib.import_module(module_name)
+            return module
+
+    def _services(protobuf_path, include_paths=None):
+        _protos(protobuf_path, include_paths)
+        with _augmented_syspath(include_paths):
+            module_name = _proto_file_to_module_name(_SERVICE_MODULE_SUFFIX,
+                                                     protobuf_path)
+            module = importlib.import_module(module_name)
+            return module
+
+    def _protos_and_services(protobuf_path, include_paths=None):
+        return (_protos(protobuf_path, include_paths=include_paths),
+                _services(protobuf_path, include_paths=include_paths))
+
+    _proto_code_cache = {}
+    _proto_code_cache_lock = threading.RLock()
+
+    class ProtoLoader(importlib.abc.Loader):
+
+        def __init__(self, suffix, codegen_fn, module_name, protobuf_path,
+                     proto_root):
+            self._suffix = suffix
+            self._codegen_fn = codegen_fn
+            self._module_name = module_name
+            self._protobuf_path = protobuf_path
+            self._proto_root = proto_root
+
+        def create_module(self, spec):
+            return None
+
+        def _generated_file_to_module_name(self, filepath):
+            components = filepath.split(os.path.sep)
+            return ".".join(
+                components[:-1] + [os.path.splitext(components[-1])[0]])
+
+        def exec_module(self, module):
+            assert module.__name__ == self._module_name
+            code = None
+            with _proto_code_cache_lock:
+                if self._module_name in _proto_code_cache:
+                    code = _proto_code_cache[self._module_name]
+                    exec(code, module.__dict__)
+                else:
+                    files = self._codegen_fn(
+                        self._protobuf_path.encode('ascii'),
+                        [path.encode('ascii') for path in sys.path])
+                    # NOTE: The files are returned in topological order of dependencies. Each
+                    # entry is guaranteed to depend only on the modules preceding it in the
+                    # list and the last entry is guaranteed to be our requested module. We
+                    # cache the code from the first invocation at module-scope so that we
+                    # don't have to regenerate code that has already been generated by protoc.
+                    for f in files[:-1]:
+                        module_name = self._generated_file_to_module_name(
+                            f[0].decode('ascii'))
+                        if module_name not in sys.modules:
+                            if module_name not in _proto_code_cache:
+                                _proto_code_cache[module_name] = f[1]
+                            importlib.import_module(module_name)
+                    exec(files[-1][1], module.__dict__)
```

Comment:
Is `exec` the expected way to load a module? Does `importlib` has some way to load the source code?

Reply:
According to everything I've read, `exec` is the conventional way to load contents into a module object.